In [1]:
import pandas as pd

In [2]:
tennis = pd.read_csv('atp_tennis.csv')

C:\Users\timst\AppData\Local\Temp\ipykernel_29852\1539982663.py:1: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  tennis = pd.read_csv('atp_tennis.csv')


In [3]:
tennis.head()

,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Odd_2,Score
0,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Dosedel S.,Ljubicic I.,Dosedel S.,63,77,-1,-1,-1.0,-1.0,6-4 6-2
1,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Clement A.,Enqvist T.,Enqvist T.,56,5,-1,-1,-1.0,-1.0,3-6 3-6
2,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Escude N.,Baccanello P.,Escude N.,40,655,-1,-1,-1.0,-1.0,6-7 7-5 6-3
3,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Knippschild J.,Federer R.,Federer R.,87,65,-1,-1,-1.0,-1.0,1-6 4-6
4,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Fromberg R.,Woodbridge T.,Fromberg R.,81,198,-1,-1,-1.0,-1.0,7-6 5-7 6-4


In [4]:
tennis.index

RangeIndex(start=0, stop=67460, step=1)

In [5]:
tennis.describe()

,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1
count,67460.000000,67460.000000,67460.000000,67460.000000,67460.000000,67460.000000
mean,3.376727,75.789594,75.533931,1138.918589,1144.835443,2.417265
std,0.782010,100.489150,100.739889,1721.413413,1750.800536,2.663592
min,3.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000
25%,3.000000,24.000000,25.000000,137.000000,138.000000,1.343458
50%,3.000000,53.000000,54.000000,705.000000,704.000000,1.749625
75%,3.000000,92.000000,92.000000,1255.000000,1255.000000,2.750000
max,5.000000,3390.000000,4915.000000,16950.000000,16950.000000,67.000000


If you set the attribute ```include``` to ```'all'```, you can additionally get information about non-numerical columns.

In [6]:
tennis.describe(include='all')

,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Odd_2,Score
count,67460,67460,67460,67460,67460,67460,67460.000000,67460,67460,67460,67460.000000,67460.000000,67460.000000,67460.000000,67460.000000,67460.0,67460
unique,268,6602,8,2,4,8,NaN,1524,1525,1186,NaN,NaN,NaN,NaN,NaN,2049.0,7151
top,Australian Open,2001-01-15,ATP250,Outdoor,Hard,1st Round,NaN,Federer R.,Federer R.,Federer R.,NaN,NaN,NaN,NaN,NaN,-1.0,6-3 6-4
freq,3275,125,17979,55439,36836,30888,NaN,668,725,1151,NaN,NaN,NaN,NaN,NaN,3780.0,1230
mean,NaN,NaN,NaN,NaN,NaN,NaN,3.376727,NaN,NaN,NaN,75.789594,75.533931,1138.918589,1144.835443,2.417265,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,0.782010,NaN,NaN,NaN,100.489150,100.739889,1721.413413,1750.800536,2.663592,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,NaN,NaN,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,NaN,NaN,24.000000,25.000000,137.000000,138.000000,1.343458,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,NaN,NaN,53.000000,54.000000,705.000000,704.000000,1.749625,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,NaN,NaN,92.000000,92.000000,1255.000000,1255.000000,2.750000,NaN,NaN


# Data Transformation

## Score Transformation

Right now, the score is stored as a string and therefore does not provide any useful information. If the score is transformed into an integer, it could be used to indicate how close or one-sided a victory was.

In [7]:
#Before the transformation give information on the string values in score column
tennis["Scorelength"] = tennis["Score"].str.len()
tennis.describe()


,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength
count,67460.000000,67460.000000,67460.000000,67460.000000,67460.000000,67460.000000,67460.000000
mean,3.376727,75.789594,75.533931,1138.918589,1144.835443,2.417265,9.437341
std,0.782010,100.489150,100.739889,1721.413413,1750.800536,2.663592,3.046301
min,3.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,3.000000
25%,3.000000,24.000000,25.000000,137.000000,138.000000,1.343458,7.000000
50%,3.000000,53.000000,54.000000,705.000000,704.000000,1.749625,7.000000
75%,3.000000,92.000000,92.000000,1255.000000,1255.000000,2.750000,11.000000
max,5.000000,3390.000000,4915.000000,16950.000000,16950.000000,67.000000,21.000000


This table shows that more than 75% of the data is less than 11 strings long, so most of the matches consist of 3 or less matches.

In [8]:
tennis20 = tennis[tennis["Scorelength"] == 19] 
tennis20.describe()


,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength
count,2364.000000,2364.000000,2364.000000,2364.000000,2364.000000,2364.000000,2364.0
mean,4.996616,69.222504,68.594755,1197.129019,1209.039763,2.476170,19.0
std,0.082217,77.400249,76.465863,1725.790187,1802.594843,2.518202,0.0
min,3.000000,1.000000,1.000000,-1.000000,-1.000000,-1.000000,19.0
25%,5.000000,21.000000,22.000000,209.750000,221.500000,1.330000,19.0
50%,5.000000,51.000000,50.000000,710.500000,723.500000,1.800000,19.0
75%,5.000000,93.000000,92.000000,1385.000000,1360.000000,2.953125,19.0
max,5.000000,1093.000000,1415.000000,16790.000000,13920.000000,34.000000,19.0


We know a score enty looks like this: "6-3 6-4": first the player 1 score of match 1, then a - and then the score of player 2. Each match is then separated by a space.

If every score is one integer long, we know how many matches were played based on the score length. 
Scorelength = (Number of matches) * 4 - 1

In [9]:
tennis["Matches"] = (tennis["Scorelength"] // 4) + 1 # For our calculation we divide the scorelenth by 4 and round down and add 1 for the one that doesn't get included on the round down divison 4
tennis19 = tennis[tennis["Scorelength"] <= 19] # this calculation on ly works for calculations that have a one digit score, so all matches larger than 19 it will not work for.
tennis19["Diff"]  = tennis19["Score"].str[0].fillna(0).astype(int) + tennis19["Score"].str[4].fillna(0).astype(int) + tennis19["Score"].str[8].fillna(0).astype(int) + tennis19["Score"].str[12].fillna(0).astype(int) + tennis19["Score"].str[16].fillna(0).astype(int) - tennis19["Score"].str[2].fillna(0).astype(int) - tennis19["Score"].str[6].fillna(0).astype(int)  - tennis19["Score"].str[10].fillna(0).astype(int) - tennis19["Score"].str[14].fillna(0).astype(int) - tennis19["Score"].str[18].fillna(0).astype(int)
tennis19.describe()
# To save the Score as an integer we calculate the difference between all the scores of player 1 subtracted by the scores of player 2. 

C:\Users\timst\AppData\Local\Temp\ipykernel_29852\3972346842.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tennis19["Diff"]  = tennis19["Score"].str[0].fillna(0).astype(int) + tennis19["Score"].str[4].fillna(0).astype(int) + tennis19["Score"].str[8].fillna(0).astype(int) + tennis19["Score"].str[12].fillna(0).astype(int) + tennis19["Score"].str[16].fillna(0).astype(int) - tennis19["Score"].str[2].fillna(0).astype(int) - tennis19["Score"].str[6].fillna(0).astype(int)  - tennis19["Score"].str[10].fillna(0).astype(int) - tennis19["Score"].str[14].fillna(0).astype(int) - tennis19["Score"].str[18].fillna(0).astype(int)


,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength,Matches,Diff
count,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000
mean,3.374148,75.804537,75.529820,1138.800187,1144.848752,2.417459,9.419848,2.604962,0.002791
std,0.779948,100.545487,100.682574,1720.680913,1750.931231,2.664260,3.016848,0.754212,5.712517
min,3.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,3.000000,1.000000,-17.000000
25%,3.000000,24.000000,25.000000,137.000000,139.000000,1.343500,7.000000,2.000000,-5.000000
50%,3.000000,53.000000,54.000000,705.000000,704.000000,1.750000,7.000000,2.000000,0.000000
75%,3.000000,92.000000,92.000000,1255.000000,1255.000000,2.750000,11.000000,3.000000,5.000000
max,5.000000,3390.000000,4915.000000,16950.000000,16950.000000,67.000000,19.000000,5.000000,17.000000


The original df has 67460 entries while the new one has 67353. So there is a difference of 107 entries (less than 1% of entries) which means it is a toleratable number of data missing.

In [10]:
BO3 = tennis[tennis["Best of"] == 3]
BO3.head()

,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Odd_2,Score,Scorelength,Matches
0,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Dosedel S.,Ljubicic I.,Dosedel S.,63,77,-1,-1,-1.0,-1.0,6-4 6-2,7,2
1,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Clement A.,Enqvist T.,Enqvist T.,56,5,-1,-1,-1.0,-1.0,3-6 3-6,7,2
2,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Escude N.,Baccanello P.,Escude N.,40,655,-1,-1,-1.0,-1.0,6-7 7-5 6-3,11,3
3,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Knippschild J.,Federer R.,Federer R.,87,65,-1,-1,-1.0,-1.0,1-6 4-6,7,2
4,Australian Hardcourt Championships,2000-01-03,International,Outdoor,Hard,1st Round,3,Fromberg R.,Woodbridge T.,Fromberg R.,81,198,-1,-1,-1.0,-1.0,7-6 5-7 6-4,11,3


## Other extreme cases
After filtering out all extreme cases with a score length greater than 19, we want to check whether there are any additional outliers that do not follow our number convention. Therefore, we examine all Best-of-3 matches that were not decided in 2 or 3 sets and do not follow the number convention (that is, matches with a score length of 7 or 11).


In [11]:
print(type(BO3.Score[0]))
BO3["Scorelength"] = BO3["Score"].str.len()
BO3.describe()

C:\Users\timst\AppData\Local\Temp\ipykernel_29852\3985933436.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  BO3["Scorelength"] = BO3["Score"].str.len()


<class 'str'>


,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength,Matches
count,54753.0,54753.000000,54753.000000,54753.000000,54753.000000,54753.000000,54753.000000,54753.000000
mean,3.0,77.298541,77.279345,1077.655014,1085.992950,2.236802,8.419027,2.354757
std,0.0,103.210325,104.897644,1596.848513,1637.282351,2.052488,1.916522,0.479131
min,3.0,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,3.000000,1.000000
25%,3.0,26.000000,26.000000,117.000000,117.000000,1.360000,7.000000,2.000000
50%,3.0,54.000000,55.000000,700.000000,700.000000,1.750000,7.000000,2.000000
75%,3.0,92.000000,92.000000,1215.000000,1210.000000,2.625000,11.000000,3.000000
max,3.0,3390.000000,4915.000000,16585.000000,16790.000000,51.000000,19.000000,5.000000


In [12]:
#BO3E = BO3[BO3["Scorelength"] < 7]

BO3E = BO3[BO3["Scorelength"] != 7]
BO3E = BO3E[BO3E["Scorelength"] != 11]

BO3E



,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Odd_2,Score,Scorelength,Matches
4253,French Open,2001-05-28,Grand Slam,Outdoor,Clay,2nd Round,3,Puerta M.,Ulihrach B.,Ulihrach B.,93,45,-1,-1,1.90,1.74425,6-2 3-6 3-6 3-6,15,4
12827,Wimbledon,2004-06-28,Grand Slam,Outdoor,Grass,4th Round,3,Spadea V.,Schalken S.,Schalken S.,30,23,-1,-1,2.25,1.571,2-6 5-7 6-3 2-6,15,4
18449,Wimbledon,2006-07-03,Grand Slam,Outdoor,Grass,4th Round,3,Ferrer D.,Hewitt L.,Hewitt L.,19,9,1330,1900,6.00,1.11,4-6 4-6 6-4 5-7,15,4
21209,Wimbledon,2007-07-04,Grand Slam,Outdoor,Grass,3rd Round,3,Soderling R.,Nadal R.,Nadal R.,28,2,965,5225,6.00,1.11,4-6 4-6 7-6 6-4 5-7,19,5
21941,Thailand Open,2007-09-28,International,Indoor,Hard,Quarterfinals,3,Karlovic I.,Berdych T.,Berdych T.,30,10,949,1925,2.62,1.44,5-7,3,1
34179,Wimbledon,2012-06-30,Grand Slam,Outdoor,Grass,3rd Round,3,Baghdatis M.,Murray A.,Murray A.,42,4,945,6980,8.00,1.08,5-7 6-3 5-7 1-6,15,4
41707,Wimbledon,2015-07-04,Grand Slam,Outdoor,Grass,3rd Round,3,Monfils G.,Simon G.,Simon G.,18,13,1885,2435,2.10,1.72,6-3 3-6 6-7 6-2 2-6,19,5
51819,Wimbledon,2019-07-06,Grand Slam,Outdoor,Grass,3rd Round,3,Evans D.,Sousa J.,Sousa J.,61,69,860,770,1.36,3.2,6-4 4-6 5-7 6-4 4-6,19,5
55469,Wimbledon,2021-07-03,Grand Slam,Outdoor,Grass,3rd Round,3,Cilic M.,Medvedev D.,Medvedev D.,37,2,1660,10280,5.00,1.16,7-6 6-3 3-6 3-6 2-6,19,5
56369,BNP Paribas Masters,2021-11-07,Masters 1000,Indoor,Hard,The Final,3,Medvedev D.,Djokovic N.,Djokovic N.,2,1,9540,10340,1.80,2.0,5-4 6-4 3-6 3-6,15,4


In [13]:
BO3E.describe()


,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength,Matches
count,10.0,10.000000,10.000000,10.000000,10.000000,10.000000,10.000000,10.000000
mean,3.0,36.000000,17.800000,1813.200000,3985.300000,3.703000,15.400000,4.100000
std,0.0,25.420901,22.414281,2782.421719,3994.810746,2.331295,4.788876,1.197219
min,3.0,2.000000,1.000000,-1.000000,-1.000000,1.360000,3.000000,1.000000
25%,3.0,21.250000,2.500000,881.250000,1052.500000,1.950000,15.000000,4.000000
50%,3.0,30.000000,9.500000,957.000000,2180.000000,2.435000,15.000000,4.000000
75%,3.0,40.750000,20.500000,1577.500000,6541.250000,5.750000,19.000000,5.000000
max,3.0,93.000000,69.000000,9540.000000,10340.000000,8.000000,19.000000,5.000000


We can see that only 10 games of best of 3 aren't decided in 2 or 3 matches. The print although shows that the gamescores are still following the number convention.

## Date transformation

For merging the win history and score history, it is necessary to transform the date into a numerical format or timestamp. Since the other table already stores the date as a number, we convert this table into the same format to enable the merge.

In [14]:
print(type(tennis19.Date[0]))
tennis19["Date"] = tennis["Date"].str.replace("-","")
tennis19["Date"] = tennis19["Date"].astype(int)

<class 'str'>


C:\Users\timst\AppData\Local\Temp\ipykernel_29852\762890696.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tennis19["Date"] = tennis["Date"].str.replace("-","")
C:\Users\timst\AppData\Local\Temp\ipykernel_29852\762890696.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tennis19["Date"] = tennis19["Date"].astype(int)


# Win and Score History

In [15]:
#
tennis19["Loser"] = tennis19["Player_1"]
tennis19["Loser"].loc[tennis19["Loser"] == tennis19["Winner"] ]  = tennis19["Player_2"]

tennis19 = tennis19.sort_values("Date")


tennis19["1WinB"] = (
    (tennis19["Winner"] == tennis19["Player_1"])
    .groupby(tennis19["Player_1"])
    .cumsum()
    .shift(fill_value=0)
)

tennis19["2WinB"] = (
    (tennis19["Winner"] == tennis19["Player_2"])
    .groupby(tennis19["Player_2"])
    .cumsum()
    .shift(fill_value=0)
)

tennis19["1LoseB"] = (
    (tennis19["Loser"] == tennis19["Player_1"])
    .groupby(tennis19["Player_1"])
    .cumsum()
    .shift(fill_value=0)
)

tennis19["2LoseB"] = (
    (tennis19["Loser"] == tennis19["Player_2"])
    .groupby(tennis19["Player_2"])
    .cumsum()
    .shift(fill_value=0)
)

#tennis19["1LoseB"] = 
#tennis19["2LoseB"] = 
#test = tennis19[["Player_1","Player_2","Winner","Loser"]]
#test

tennis19.describe()

C:\Users\timst\AppData\Local\Temp\ipykernel_29852\2064722105.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tennis19["Loser"] = tennis19["Player_1"]
C:\Users\timst\AppData\Local\Temp\ipykernel_29852\2064722105.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tennis19["Loser"].loc[tennis19["Loser"] == tennis19["Winner"] ]  = tennis19["Player_2"]


,Date,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength,Matches,Diff,1WinB,2WinB,1LoseB,2LoseB
count,6.735300e+04,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000
mean,2.012273e+07,3.374148,75.804537,75.529820,1138.800187,1144.848752,2.417459,9.419848,2.604962,0.002791,57.396939,57.843496,42.338174,42.175434
std,7.672246e+04,0.779948,100.545487,100.682574,1720.680913,1750.931231,2.664260,3.016848,0.754212,5.712517,70.903276,72.716126,36.957415,36.891250
min,2.000010e+07,3.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,3.000000,1.000000,-17.000000,0.000000,0.000000,0.000000,0.000000
25%,2.006010e+07,3.000000,24.000000,25.000000,137.000000,139.000000,1.343500,7.000000,2.000000,-5.000000,10.000000,10.000000,13.000000,13.000000
50%,2.012051e+07,3.000000,53.000000,54.000000,705.000000,704.000000,1.750000,7.000000,2.000000,0.000000,32.000000,32.000000,33.000000,32.000000
75%,2.019021e+07,3.000000,92.000000,92.000000,1255.000000,1255.000000,2.750000,11.000000,3.000000,5.000000,79.000000,78.000000,63.000000,62.000000
max,2.026040e+07,5.000000,3390.000000,4915.000000,16950.000000,16950.000000,67.000000,19.000000,5.000000,17.000000,559.000000,590.000000,241.000000,231.000000


## Last 10 Wins and Losses

In [17]:
tennis19[["Date","Player_1", "Player_2","Loser","Winner"]]

,Date,Player_1,Player_2,Loser,Winner
0,20000103,Dosedel S.,Ljubicic I.,Ljubicic I.,Dosedel S.
66,20000103,El Aynaoui Y.,Dupuis A.,Dupuis A.,El Aynaoui Y.
65,20000103,Arazi H.,Cherkasov A.,Arazi H.,Cherkasov A.
64,20000103,Caratti C.,Heuberger I.,Heuberger I.,Caratti C.
63,20000103,Al-Alawi S.K.,Berasategui A.,Al-Alawi S.K.,Berasategui A.
...,...,...,...,...,...
67431,20260404,Popyrin A.,Tiafoe F.,Popyrin A.,Tiafoe F.
67432,20260404,Burruchaga R.,Tirante T.A.,Tirante T.A.,Burruchaga R.
67407,20260405,Merida Aguilar D.,Navone M.,Merida Aguilar D.,Navone M.
67434,20260405,Paul T.,Burruchaga R.,Burruchaga R.,Paul T.


In [18]:
import pandas as pd

tennis19 = tennis19.sort_values("Date").reset_index(drop=True)
tennis19["match_id"] = tennis19.index

p1 = tennis19[["match_id", "Date", "Player_1", "Winner", "Loser"]].copy()
p1["Player"] = p1["Player_1"]
p1["Win"] = (p1["Winner"] == p1["Player"]).astype(int)
p1["Los"] = (p1["Loser"] == p1["Player"]).astype(int)

p2 = tennis19[["match_id", "Date", "Player_2", "Winner", "Loser"]].copy()
p2["Player"] = p2["Player_2"]
p2["Win"] = (p2["Winner"] == p2["Player"]).astype(int)
p2["Los"] = (p2["Loser"] == p2["Player"]).astype(int)


history = pd.concat([
    p1[["match_id", "Date", "Player", "Win", "Los"]],
    p2[["match_id", "Date", "Player", "Win", "Los"]]
])

history = history.sort_values(["Player", "Date", "match_id"])

history["WinsLast10"] = (
    history.groupby("Player")["Win"]
    .transform(lambda x: x.shift().rolling(10, min_periods=1).sum())
    .fillna(0)
)

history["LosLast10"] = (
    history.groupby("Player")["Los"]
    .transform(lambda x: x.shift().rolling(10, min_periods=1).sum())
    .fillna(0)
)

In [19]:
history.head(50)

,match_id,Date,Player,Win,Los,WinsLast10,LosLast10
19184,19184,20060926,Hajek J.,1,0,0.0,0.0
19219,19219,20060928,Hajek J.,0,1,1.0,0.0
14032,14032,20050104,Abdulla M.,0,1,0.0,0.0
8044,8044,20020909,Abel M.,0,1,0.0,0.0
8199,8199,20020930,Abel M.,0,1,0.0,1.0
9034,9034,20030227,Abel M.,0,1,0.0,2.0
9828,9828,20030610,Abel M.,0,1,0.0,3.0
10169,10169,20030714,Abel M.,1,0,0.0,4.0
10203,10203,20030716,Abel M.,0,1,1.0,4.0
10615,10615,20030825,Abel M.,0,1,1.0,5.0


In [20]:
history[history["Player"] == "Dosedel S."].head(50)

,match_id,Date,Player,Win,Los,WinsLast10,LosLast10
0,0,20000103,Dosedel S.,1,0,0.0,0.0
17,17,20000103,Dosedel S.,0,1,1.0,0.0
131,131,20000110,Dosedel S.,1,0,1.0,1.0
137,137,20000110,Dosedel S.,1,0,2.0,1.0
150,150,20000110,Dosedel S.,0,1,3.0,1.0
153,153,20000117,Dosedel S.,1,0,3.0,2.0
218,218,20000117,Dosedel S.,1,0,4.0,2.0
245,245,20000117,Dosedel S.,1,0,5.0,2.0
268,268,20000117,Dosedel S.,0,1,6.0,2.0
624,624,20000306,Dosedel S.,0,1,6.0,3.0


In [21]:
tennis19

,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,...,Score,Scorelength,Matches,Diff,Loser,1WinB,2WinB,1LoseB,2LoseB,match_id
0,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Dosedel S.,Ljubicic I.,Dosedel S.,...,6-4 6-2,7,2,6,Ljubicic I.,0,0,0,0,0
1,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,2nd Round,3,Hewitt L.,Petrovic D.,Hewitt L.,...,6-0 6-2,7,2,10,Petrovic D.,0,1,1,0,1
2,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Clement A.,Enqvist T.,Enqvist T.,...,3-6 3-6,7,2,-6,Clement A.,1,0,0,1,2
3,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Escude N.,Baccanello P.,Escude N.,...,6-7 7-5 6-3,11,3,4,Baccanello P.,0,1,1,1,3
4,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Knippschild J.,Federer R.,Federer R.,...,1-6 4-6,7,2,-7,Knippschild J.,1,0,0,1,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
67348,U.S.Men's Clay Court Championships,20260404,ATP250,Outdoor,Clay,Semifinals,3,Tiafoe F.,Paul T.,Paul T.,...,5-7 6-4 6-7,11,3,-1,Tiafoe F.,54,68,62,81,67348
67349,Tiriac Open,20260404,ATP250,Outdoor,Clay,Semifinals,3,Van De Zandschulp B.,Navone M.,Navone M.,...,7-5 6-7 5-7,11,3,-1,Van De Zandschulp B.,6,15,4,25,67349
67350,U.S.Men's Clay Court Championships,20260405,ATP250,Outdoor,Clay,The Final,3,Paul T.,Burruchaga R.,Paul T.,...,6-1 3-6 7-5,11,3,4,Burruchaga R.,2,18,2,23,67350
67351,Tiriac Open,20260405,ATP250,Outdoor,Clay,The Final,3,Merida Aguilar D.,Navone M.,Navone M.,...,2-6 6-4 5-7,11,3,-4,Merida Aguilar D.,6,10,4,12,67351


In [22]:
tennis19.describe()

,Date,Best of,Rank_1,Rank_2,Pts_1,Pts_2,Odd_1,Scorelength,Matches,Diff,1WinB,2WinB,1LoseB,2LoseB,match_id
count,6.735300e+04,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000,67353.000000
mean,2.012273e+07,3.374148,75.804537,75.529820,1138.800187,1144.848752,2.417459,9.419848,2.604962,0.002791,57.396939,57.843496,42.338174,42.175434,33676.000000
std,7.672246e+04,0.779948,100.545487,100.682574,1720.680913,1750.931231,2.664260,3.016848,0.754212,5.712517,70.903276,72.716126,36.957415,36.891250,19443.280677
min,2.000010e+07,3.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,3.000000,1.000000,-17.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2.006010e+07,3.000000,24.000000,25.000000,137.000000,139.000000,1.343500,7.000000,2.000000,-5.000000,10.000000,10.000000,13.000000,13.000000,16838.000000
50%,2.012051e+07,3.000000,53.000000,54.000000,705.000000,704.000000,1.750000,7.000000,2.000000,0.000000,32.000000,32.000000,33.000000,32.000000,33676.000000
75%,2.019021e+07,3.000000,92.000000,92.000000,1255.000000,1255.000000,2.750000,11.000000,3.000000,5.000000,79.000000,78.000000,63.000000,62.000000,50514.000000
max,2.026040e+07,5.000000,3390.000000,4915.000000,16950.000000,16950.000000,67.000000,19.000000,5.000000,17.000000,559.000000,590.000000,241.000000,231.000000,67352.000000


# Publish

In [23]:
import numpy as np

In [24]:
tennis19["WinRank"] = np.where(
    tennis19["Player_1"] == tennis19["Winner"],
    tennis19["Rank_1"],
    tennis19["Rank_2"]
)

tennis19["LosRank"] = np.where(
    tennis19["Player_1"] == tennis19["Winner"],
    tennis19["Rank_2"],
    tennis19["Rank_1"]
)

tennis19

,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,...,Matches,Diff,Loser,1WinB,2WinB,1LoseB,2LoseB,match_id,WinRank,LosRank
0,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Dosedel S.,Ljubicic I.,Dosedel S.,...,2,6,Ljubicic I.,0,0,0,0,0,63,77
1,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,2nd Round,3,Hewitt L.,Petrovic D.,Hewitt L.,...,2,10,Petrovic D.,0,1,1,0,1,24,208
2,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Clement A.,Enqvist T.,Enqvist T.,...,2,-6,Clement A.,1,0,0,1,2,5,56
3,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Escude N.,Baccanello P.,Escude N.,...,3,4,Baccanello P.,0,1,1,1,3,40,655
4,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Knippschild J.,Federer R.,Federer R.,...,2,-7,Knippschild J.,1,0,0,1,4,65,87
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
67348,U.S.Men's Clay Court Championships,20260404,ATP250,Outdoor,Clay,Semifinals,3,Tiafoe F.,Paul T.,Paul T.,...,3,-1,Tiafoe F.,54,68,62,81,67348,21,18
67349,Tiriac Open,20260404,ATP250,Outdoor,Clay,Semifinals,3,Van De Zandschulp B.,Navone M.,Navone M.,...,3,-1,Van De Zandschulp B.,6,15,4,25,67349,60,62
67350,U.S.Men's Clay Court Championships,20260405,ATP250,Outdoor,Clay,The Final,3,Paul T.,Burruchaga R.,Paul T.,...,3,4,Burruchaga R.,2,18,2,23,67350,21,77
67351,Tiriac Open,20260405,ATP250,Outdoor,Clay,The Final,3,Merida Aguilar D.,Navone M.,Navone M.,...,3,-4,Merida Aguilar D.,6,10,4,12,67351,60,136


In [25]:
tennis19.to_csv("BearbTennis.csv", index=False)

In [26]:
tennis20000103 = tennis19[tennis19["Date"]==20000103 ]
tennis20000103

,Tournament,Date,Series,Court,Surface,Round,Best of,Player_1,Player_2,Winner,...,Matches,Diff,Loser,1WinB,2WinB,1LoseB,2LoseB,match_id,WinRank,LosRank
0,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Dosedel S.,Ljubicic I.,Dosedel S.,...,2,6,Ljubicic I.,0,0,0,0,0,63,77
1,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,2nd Round,3,Hewitt L.,Petrovic D.,Hewitt L.,...,2,10,Petrovic D.,0,1,1,0,1,24,208
2,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Clement A.,Enqvist T.,Enqvist T.,...,2,-6,Clement A.,1,0,0,1,2,5,56
3,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Escude N.,Baccanello P.,Escude N.,...,3,4,Baccanello P.,0,1,1,1,3,40,655
4,Australian Hardcourt Championships,20000103,International,Outdoor,Hard,1st Round,3,Knippschild J.,Federer R.,Federer R.,...,2,-7,Knippschild J.,1,0,0,1,4,65,87
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
85,Qatar Open,20000103,International,Outdoor,Hard,Quarterfinals,3,Santoro F.,Bastl G.,Santoro F.,...,2,6,Bastl G.,0,2,1,0,85,32,85
86,Qatar Open,20000103,International,Outdoor,Hard,Quarterfinals,3,Mirnyi M.,Schuettler R.,Schuettler R.,...,2,-7,Mirnyi M.,2,0,0,1,86,47,74
87,Qatar Open,20000103,International,Outdoor,Hard,Semifinals,3,Santoro F.,Kiefer N.,Santoro F.,...,2,4,Kiefer N.,1,1,1,0,87,32,6
88,Qatar Open,20000103,International,Outdoor,Hard,2nd Round,3,Squillari F.,Schalken S.,Schalken S.,...,3,-8,Squillari F.,1,0,0,1,88,44,49
